### PR TITLE
SQL Prepared Statement

### PR DESCRIPTION
**Pull request by fedefernandez**
Adds the prepared statement implementation

Also removes the unnecessary Java files

Please @juanpedromoreno could you review? Thanks

### CODE REVIEW

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala [hunk 1]
```diff
@@ -22,15 +26,17 @@ class SQLDroidStatement(
...
-  private[this] var maxRows: Option[Int] = None
+  protected var connection: Option[SQLDroidConnection] = Option(sqlDroidConnection)
 
-  private[this] var resultSet: Option[ResultSet] = None
+  protected var maxRows: Option[Int] = None
 
-  private[this] var updateCount: Option[Int] = None
+  protected var resultSet: Option[ResultSet] = None
+
+  protected var updateCount: Option[Int] = None
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala:39]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582811/issues/source?bid=2442814&fileBranchId=2752104#l39)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala [hunk 2]
```diff
@@ -22,15 +26,17 @@ class SQLDroidStatement(
...
+  protected val batchList = new mutable.MutableList[String]
 
-  private[this] var maxRows: Option[Int] = None
+  protected var connection: Option[SQLDroidConnection] = Option(sqlDroidConnection)
 
-  private[this] var resultSet: Option[ResultSet] = None
+  protected var maxRows: Option[Int] = None
 
-  private[this] var updateCount: Option[Int] = None
+  protected var resultSet: Option[ResultSet] = None
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala:37]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582811/issues/source?bid=2442814&fileBranchId=2752104#l37)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/StatementArguments.scala [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  def toArray: scala.Array[AnyRef] = scala.Array.empty
+
+}
+
+class PreparedStatementArguments extends StatementArguments {
+
+  private[this] var argumentsList: NonEmptyList[mutable.Map[Int, AnyRef]] = NonEmptyList(mutable.Map.empty)
+
+  // Prepared statements count from 1
+  private[this] var maxIndex: Int = 1
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/StatementArguments.scala:22]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582812/issues/source?bid=2442814&fileBranchId=2752104#l22)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala [hunk 4]
```diff
@@ -22,15 +26,17 @@ class SQLDroidStatement(
...
+  val defaultUpdateCount = -1
 
-  private[this] var connection: Option[SQLDroidConnection] = Option(sqlDroidConnection)
+  protected val batchList = new mutable.MutableList[String]
 
-  private[this] var maxRows: Option[Int] = None
+  protected var connection: Option[SQLDroidConnection] = Option(sqlDroidConnection)
 
-  private[this] var resultSet: Option[ResultSet] = None
+  protected var maxRows: Option[Int] = None
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala:35]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582811/issues/source?bid=2442814&fileBranchId=2752104#l35)

File: src/test/scala/com/fortysevendeg/android/sqlite/statement/StatementArgumentsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+import com.fortysevendeg.android.sqlite._
+import org.specs2.mock.Mockito
+import org.specs2.mutable.Specification
+import org.specs2.specification.Scope
+
+trait StatementArgumentsSpecification
+  extends Specification
+  with Mockito {
+
+  trait StatementArgumentsScope
```
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/statement/StatementArgumentsSpec.scala]: Codacy Issue found: [Redundant braces after class definition](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582835/issues/source?bid=2442814&fileBranchId=2752104#l14)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatement.scala [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+import java.util.Calendar
+
+import com.fortysevendeg.android.sqlite.{SQLDroidConnection, _}
+import com.fortysevendeg.android.sqlite.logging.{AndroidLogWrapper, LogWrapper}
+import com.fortysevendeg.android.sqlite.util.StreamUtils._
+import StatementInfo._
+
+import scala.util.{Failure, Success, Try}
+
+class SQLDroidPreparedStatement(
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatement.scala:16]: Codacy Issue found: [Number of methods in class exceeds 30](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582810/issues/source?bid=2442814&fileBranchId=2752104#l16)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala [hunk 7]
```diff
@@ -22,15 +26,17 @@ class SQLDroidStatement(
...
   val selectLastRowId = "SELECT last_insert_rowid()"
 
-  private[this] val batchList = new mutable.MutableList[String]
+  val defaultUpdateCount = -1
 
-  private[this] var connection: Option[SQLDroidConnection] = Option(sqlDroidConnection)
+  protected val batchList = new mutable.MutableList[String]
 
-  private[this] var maxRows: Option[Int] = None
+  protected var connection: Option[SQLDroidConnection] = Option(sqlDroidConnection)
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatement.scala:33]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582811/issues/source?bid=2442814&fileBranchId=2752104#l33)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/StatementArguments.scala [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+  def toStringArray: scala.Array[String] = scala.Array.empty
+
+  def toArray: scala.Array[AnyRef] = scala.Array.empty
+
+}
+
+class PreparedStatementArguments extends StatementArguments {
+
+  private[this] var argumentsList: NonEmptyList[mutable.Map[Int, AnyRef]] = NonEmptyList(mutable.Map.empty)
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/StatementArguments.scala:19]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582812/issues/source?bid=2442814&fileBranchId=2752104#l19)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/StatementArguments.scala [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+      setArgument(position, a.toLong)
+    }
+
+  private[this] def addToHead(position: Int, arg: AnyRef) = {
+    argumentsList.head += (position -> arg)
+    if (position > maxIndex) maxIndex = position
+  }
+
+  private[this] def setNullArgument(position: Int): Unit = {
+    argumentsList.head.remove(position)
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/StatementArguments.scala]: Codacy Issue found: [Usage of head/last on collections.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582812/issues/source?bid=2442814&fileBranchId=2752104#l124)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatement.scala [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+  override def executeQuery(): ResultSet =
+    executeQueryWithArgs(sql, Some(arguments))
+
+  override def executeUpdate(): Int =
+    executeUpdateWithArgs(sql, Some(arguments))
+
+  override def executeBatch(): scala.Array[Int] = withOpenConnection { db =>
+    val updateArray = arguments map { array =>
+      Try(db.execSQL(sql, array)) match {
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatement.scala:80]: Codacy Issue found: [Default case not specified.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/821582810/issues/source?bid=2442814&fileBranchId=2752104#l80)

File: src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatement.scala [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        case Success(_) =>
+          if (isChange(sql)) db.changedRowCount()
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatement.scala]: Maybe you could put the if conditional in the pattern matching `case`

File: src/test/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatementSpec.scala [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+    val connection: SQLDroidConnection = new SQLDroidConnection(databaseName = "databaseName", logWrapper = new TestLogWrapper) {
+      override protected def createDatabase(): Option[SQLDroidDatabase] = Some(database)
+    }
+
+    val arguments = mock[PreparedStatementArguments]
+  }
+
+  trait WithoutConnection extends SQLDroidPreparedStatementScope {
+
+    val columnGenerated = None
```
Comment (juanpedromoreno) [inline src/test/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidPreparedStatementSpec.scala]: This val is not being used.

File: src/test/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatementSpec.scala [hunk 13]
```diff
@@ -153,7 +153,7 @@ class SQLDroidStatementSpec
     }
 
     "return an array with the results of changedRowCount and the sum when the bacth has elements" in new WithoutColumnGenerated {
-      val batch = Seq(selectSql, selectSql, selectSql)
+      val batch = Seq(insertSql, insertSql, insertSql)
```
Comment (juanpedromoreno) [inline src/test/scala/com/fortysevendeg/android/sqlite/statement/SQLDroidStatementSpec.scala]: Seq.fill(3)(insertSql)

File: src/test/scala/com/fortysevendeg/android/sqlite/statement/StatementArgumentsSpec.scala [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    extends Scope {
+
+  }
```
Comment (juanpedromoreno) [inline src/test/scala/com/fortysevendeg/android/sqlite/statement/StatementArgumentsSpec.scala]: As Codacy said, you can just define the trait as:

`trait StatementArgumentsScope extends Scope`

**Comment by juanpedromoreno** [comment]
Minor comments add and LGTM! Good job @fedefernandez !

giphy

**Comment by fedefernandez** [comment]
:joy: thanks!

**Pull request merged**